src/bsp/platform/cc3220sf/xi_bsp_fwu_cc3220sf.c

```diff
@@ -0,0 +1,143 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_bsp_fwu.h>
+#include <xi_bsp_debug.h>
+#include <xively_error.h>
+
+#include "simplelink.h"
+#include "hw_types.h"
+#include "prcm.h"
+
+typedef struct xi_bsp_fwu_cc3220sf_boot_info_t
+{
+  _u8 active_image;
+  _u32 img_status;
+  _u32 start_watchdog_key;
+  _u32 start_watchdog_time;
+}xi_bsp_fwu_cc3220sf_boot_info_t;
+
+static void _reboot_device()
+{
+	_i16 status;
+	
+	xi_bsp_debug_logger(" Rebooting the CC3220!" );	
+    status = sl_Stop(200);
+	
+	if ( 0 > status )
+	{
+		xi_bsp_debug_format( "SimpleLink processor failed to stop with error: [ %i ]", status );
+		
+		while ( 1 );
+	}
+	
+	/* Reset the MCU in order to test the bundle */
+    PRCMHibernateCycleTrigger();
+}
+
+static _i16 _ota_get_pending_commit()
+{
+	SlFsControlGetStorageInfoResponse_t fs_control_storage_response;
+    _i16 status;
+
+    /* read bundle state and check if is "PENDING COMMIT" */
+    status = ( _i16 )sl_FsCtl( ( SlFsCtl_e)SL_FS_CTL_GET_STORAGE_INFO, 
+	                                0, NULL , NULL , 0, ( uint8_t * )&fs_control_storage_response, 
+									sizeof( SlFsControlGetStorageInfoResponse_t ), NULL );
+    if ( 0 > status )
+    {
+        xi_bsp_debug_format( "Error checking pending commit, status: [ %i ]", status );
+        return status;
+    }
+
+    return ( SL_FS_BUNDLE_STATE_PENDING_COMMIT == 
+	       ( SlFsBundleState_e )fs_control_storage_response.FilesUsage.Bundlestate );
+}
+
+static _i16 _ota_commit()
+{
+	 SlFsControl_t fs_control;
+    int16_t status;
+
+    fs_control.IncludeFilters = 0;
+    status = ( _i16 )sl_FsCtl( SL_FS_CTL_BUNDLE_COMMIT, 0, NULL ,(uint8_t *)&fs_control, 
+	                            sizeof( SlFsControl_t ), NULL, 0 , NULL );
+								
+    if( 0 > status )
+    {
+        xi_bsp_debug_format( "Error attempting a commit, status: [ %i ]", status );
+    }
+    return status;
+}
+
+static _i16 _ota_rollback()
+{
+	SlFsControl_t fs_control;
+    _i16 status;
+
+    fs_control.IncludeFilters = 0;
+    status = ( _i16 )sl_FsCtl( SL_FS_CTL_BUNDLE_ROLLBACK, 0, NULL ,(uint8_t *)&fs_control, 
+	                            sizeof( SlFsControl_t ), NULL, 0 , NULL );
+	
+    if( 0 > status )
+    {
+        xi_bsp_debug_format( "Error attempting a rollback, status: [ %i ]", status );
+    }
+    return status;	
+}
+
+
+
+uint8_t xi_bsp_fwu_is_this_firmware( const char* const resource_name )
+{
+    return ( 0 == strcmp( "firmware.tar", resource_name ) ) ? 1 : 0;
+}
+
+xi_bsp_fwu_state_t xi_bsp_fwu_on_new_firmware_ok()
+{
+    if ( _ota_get_pending_commit() )
+    {
+		xi_bsp_debug_logger(" Pending commit, firmware OK." );
+        
+		/* What should we do if we get a failure to commit? Reboot? */
```

Comment:
hm, great question. commit failure means the device couldn't set the current FW to the default one, which means after next reboot not this but the previous FW will start. Right?! If we rebooted here it'd just shorten this new FW's runtime and fallback immediately to previous FW.
I don't know, should we retry _ota_commit() later? Schedule a job for this? From this BSP function the client's scheduler API isn't accessible though.

Reply:
For the CC3220SF there is a bootloader watchdog timer that is set when the firmware tar file is being accessed only. If the commit fails I believe that this watchdog timer will elapse and force a rollback. Does the watchdog timer need to be set on all updates including ones that do not update the firmware?